formats/protobuf/commonTest/src/kotlinx/serialization/protobuf/ProtobufOneOfTest.kt

```diff
@@ -0,0 +1,464 @@
+    sealed interface InnerOneOf
+
+    @Serializable
+    @ProtoNumber(1)
+    data class InnerInt(val i: Int) : InnerOneOf
+
+    @Serializable
+    @ProtoNumber(2)
+    data class InnerString(val s: String) : InnerOneOf
+
+    @Test
+    fun testEncodeNestedOneOf() {
+        val data = OneOfData(NestedOneOfType(i = InnerNested(InnerInt(32))), "foo")
+        ProtoBuf.encodeToHexString(OneOfData.serializer(), data).also {
+            println(it)
+            /**
+             * 9: {1: 32}
+             * 3: {"foo"}
+             */
+            assertEquals("4a0208201a03666f6f", it)
+        }
+    }
+
+    @Test
+    fun testDecodeNestedOneOf() {
+        val data = OneOfData(NestedOneOfType(i = InnerNested(InnerInt(32))), "foo")
+        ProtoBuf.decodeFromHexString<OneOfData>("4a0208201a03666f6f").also {
+            println(it)
+            /**
+             * 9: {1: 32}
+             * 3: {"foo"}
+             */
+            assertEquals(data, it)
+        }
+    }
+
+    @Serializable
+    data class DoubleOneOfElement(
+        @ProtoOneOf val one: IType,
+        @ProtoNumber(3) val name: String,
+        @ProtoOneOf val two: OtherType
+    )
+
+    interface OtherType
+
+    @Serializable
+    @ProtoNumber(4)
+    data class OtherIntType(val i: Int) : OtherType
+
+    @Serializable
+    @ProtoNumber(5)
+    data class OtherStringType(val s: String) : OtherType
+
+    @Test
+    fun testEncodeDoubleOneOf() {
+        val module = SerializersModule {
+            polymorphic(OtherType::class) {
+                subclass(OtherStringType::class)
+            }
+        }
+        val buf = ProtoBuf {
+            serializersModule = module
+        }
+        val data = DoubleOneOfElement(
+            IntType(32),
+            "foo",
+            OtherStringType("bar")
+        )
+        buf.encodeToHexString(DoubleOneOfElement.serializer(), data).also {
+            println(it)
+            assertEquals("08201a03666f6f2a03626172", it)
+        }
+
+        assertFailsWith<SerializationException> {
+            buf.encodeToHexString(
+                DoubleOneOfElement.serializer(), DoubleOneOfElement(
+                    IntType(32),
+                    "foo",
+                    OtherIntType(32)
+                )
+            )
+        }
+    }
+
+    @Test
+    fun testDecodeDoubleOneOf() {
+        val module = SerializersModule {
+            polymorphic(OtherType::class) {
+                subclass(OtherStringType::class)
+            }
+        }
+        val buf = ProtoBuf {
+            serializersModule = module
+        }
+        val data = DoubleOneOfElement(
+            IntType(32),
+            "foo",
+            OtherStringType("bar")
+        )
+        buf.decodeFromHexString<DoubleOneOfElement>("08201a03666f6f2a03626172").also {
+            assertEquals(data, it)
+        }
+
+        assertFailsWith<SerializationException> {
+            buf.decodeFromHexString<DoubleOneOfElement>("082018666f6f2020")
```

Comment:
Please drop a comment explaining what is in the input and why it should considered a failure. Currently it throws `Expected wire type 2, but found 0` which is not very helpful (it is already existing exception, so that's one on me).

Reply:
Sadly I cannot retell what this check is for. The hex data seems just broken. I will make it correct ( with other test change requests ) in few days.